Dynawaltz integration

nicolas-pierr opened the pull request:
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Licence headers

**Does this PR already have an issue describing the problem ?**
Fixes #506

Commit:
dynawaltz dependency

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
added dynawaltz context, dynamic model mapping

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
C function export

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
start stop time of simulation

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
basic event and timeseries handler

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
context taking all parameters

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

nicolas-pierr commented:
Hello @geofjamg and @flo-dup 
I'm basing the dynawaltz integration on the powsybl-dynawo api.
In the given part of the example, you want to add dynamically events with their timestep : 
```
simu = pp.dynamic.Simulation()
simu.add_event(10, pp.dynamic.EventType.BRANCH_DISCONNECTION, 'l1')
simu.add_event(15, pp.dynamic.EventType.SWITCH_CLOSE, 's1')
```
However, powsybl-dynawo doesn't allow it, it is only able to load via a parameter file (model.par).

Do you want this feature to be added or should it work like the current powsybl-dynawo way and load configurations and parameters from files ?

Also does this conversation should be in the issue or in this PR ?

Thank you

Commit:
unused import

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
formatting

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
using docatch in java CEntrypoint and fixing signature

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

flo-dup commented:
Indeed the timestep is set through the `.par` file. Therefore you need to modify that file to put the timestep defined by the user. I think this needs to be done in powsybl-dynawo though, because it makes sense for any user to define the timestep in the event and not among some values in the `.par` file. So far the `.par` file is seen as an input or a config file in powsybl-dynawo, we need to change this...

flo-dup commented:
We could change the API in the meantime (I mean before powsybl-dynawo is modified to enable time step modifications);
```python
simu = pp.dynamic.Simulation()
simu.add_event(parametersSetId1, pp.dynamic.EventType.BRANCH_DISCONNECTION, 'l1')
simu.add_event(parametersSetId2, pp.dynamic.EventType.SWITCH_CLOSE, 's1')
```

sylvlecl commented on java/src/main/java/com/powsybl/python/dynamic/DynamicModelMapper.java:
```diff
@@ -0,0 +1,71 @@
+package com.powsybl.python.dynamic;
```
Don't forget to add license headers at some point ;)

EtienneLt replied:
and SPDX in the copyright.
```java
 /**
 * Copyright (c) 2022, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *  SPDX-License-Identifier: MPL-2.0
 */
```

Commit:
added Current limit automaton

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
renaming Timeseries to Curve to match powsybl dynawo

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
typo + added all classic mapping

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
licence headers

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
change handle of String to cpp

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
c++ and python api

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
better logging and provider interface

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
release guard to prevent deadlock on computing thread

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
remove debug verbosity

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
fix python typing

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
fix python linting

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
Adds result status to DynamicResults

Signed-off-by: Jean-Luc Bouchot <jean-luc.bouchot@artelys.com>

jlbouchot committed:
Adds return type for DynResult Status

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

Commit:
curve printing from results

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
more licence

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
add multiples curves

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

jlbouchot committed:
Adds generic add_event function

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

jlbouchot committed:
Corrects from CI and linting

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

jlbouchot committed:
Adds forgotten return type

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

jlbouchot committed:
Solves import orders

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

jlbouchot committed:
Deletes superfluous empty lines

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

jlbouchot committed:
Checks all pylint issues

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

sylvlecl commented on cpp/src/bindings.cpp:
```diff
@@ -640,4 +640,43 @@ PYBIND11_MODULE(_pypowsybl, m) {
 
     m.def("close", &pypowsybl::closePypowsybl, "Closes pypowsybl module.");
+    /************************DYNAMIC SIMULATION WITH DYNAWALTZ*******************************/
+    //TODO descriptions of methods
```
Actually docstrings here are not important, they are not exposed to the user.

About the separation comment above : I would propose to move everything to a dedicated function instead. We should probably have done it before to better organize the bindings.

sylvlecl commented on cpp/src/bindings.cpp:
```diff
@@ -640,4 +640,43 @@ PYBIND11_MODULE(_pypowsybl, m) {
+        py::arg("dynamic_model"), py::arg("network"), py::arg("dynamic_mapping"), py::arg("event_mapping"), py::arg("timeseries_mapping"), py::arg("start"), py::arg("stop"));
+
+    //model mapping
+    m.def("add_alphabeta_load", &pypowsybl::addAlphaBetaLoad, "TODO desc", py::arg("dynamic_mapping_handle"), py::arg("static_id"), py::arg("dynamic_param"));
```
I think instead of having multiple methods here, we could have only one for all models, which would receive a dataframe input exactly as we have done for all updates methods of the network. 
This will allow to:
- minimize the code boiler plate necessary for each kind of event
- map multiple models in only one call. This is very important for performances, since calling 100s of methods in a for loop will have very poor performance in python.

sylvlecl replied:
(but of course we still need to have multiple methods on the python side, for the user api

nicolas-pierr replied:
The change has been done here : f90ea09cb29b308e0b84299496bc1948c0bb7bff .
However some mappings like CurrentLimitAutomaton (and maybe other mapping in the future) need more infos.

The actual behavior is to have some default values (possible because CurrentLimitAutomaton needs an enum argument) to be able to map them _via_ the dataframe call.
But they still need a special python/java binding to be able to specify all the arguments.

Commit:
moving python bindings to function

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
moving dataframe utils out of networkCFunction

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
dynamic mapping via dataframe

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
utility to add default index to kwargs

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
doc and add kwargs support

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

jlbouchot committed:
Starts dynamic unit tests

Signed-off-by: Jean-Luc Bouchot <jlbouchot@gmail.com>

Commit:
doc

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
typing and formatting

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
minimum tests

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

sylvlecl commented on pypowsybl/dynamic.py:
```diff
@@ -0,0 +1,190 @@
+        _pp.add_current_limit_automaton(
+            self._handle, static_id, dynamic_param, branch_side.value)
+
+    def add_all_dynamic_mappings(self, df: _pd.DataFrame = None, **kwargs: _Union[str, DynamicMappingType]) -> None:
```
This is not exactly the design I had in mind: we should have the same approach as in the network class:
- all methods above should accept a dataframe or keywords as arguments (`add_current_limit_automaton` etc), to be able to add at once multiple instances of the model
- this method should be removed: we don't want to support mixing different models in the same dataframe, because as you underlined, they may have different arguments, hence different columns, which this approach forbids.

sylvlecl commented on java/src/main/java/com/powsybl/python/dynamic/DynamicModelMapper.java:
```diff
@@ -0,0 +1,136 @@
+            DynamicMappingType type = DynamicMappingType.fromCValue(dynamicModelSeries.get(row));
+            String staticId = staticIdSeries.get(row);
+            String parameterSetId = parameterSetIdSeries.get(row);
+            switch (type) {
```
Instead of this switch/case, we should imitate the facilities already in place for the network, with the `NetworkElementAdder` interface.

This should allow to have different columns for the current limit automaton, in particular.

nicolas-pierr replied:
I have rewritten the whole part about dynamic mapping to work with adders like in networks.

sylvlecl commented on pypowsybl/dynamic.py:
```diff
@@ -0,0 +1,190 @@
+
+
+class EventType(_Enum):
+    QUADRIPOLE_DISCONNECT = 'QUADRIPOLE_DISCONNECT'
```
We really need to re-work the naming of the events:
- I don't think "quadripole" is an english word ? In any case, in powsybl, we talk about "branch" instead
- I am not sure what "set point boolean" means: does it change the regulation status of a generator ?
- BRANCH_DISCONNECTION : is it a duplicate of QUADRIPOLE_DISCONNECT?  :)

nicolas-pierr replied:
From what I understood in powsybl-dynawaltz package, "set point boolean" events can only be used with a generator id. It allows to connect/disconnect the generator.
I'll rename quadripole to branch in pypowsybl, but powsybl-dynawaltz still use "quadripole".
And yes now that the naming is clearer, BRANCH_DISCONNECTION and QUADRIPOLE_DISCONNECT are the same event (and refer to the QUADRIPOLE_DISCONNECT  event in powsybl-dynawaltz)

sylvlecl commented on pypowsybl/dynamic.py:
```diff
@@ -0,0 +1,190 @@
+    def __init__(self, handle: _pp.JavaHandle) -> None:
+        self._handle = handle
+
+    def state(self) -> str:
```
let's keep only `status`

sylvlecl commented on pypowsybl/dynamic.py:
```diff
@@ -0,0 +1,190 @@
+
+class SimulationResult:
+    def __init__(self, handle: _pp.JavaHandle) -> None:
+        self._handle = handle
```
I propose to initialize all python data from the java handle (status, curves) in the constructor, this will avoid to call the java part each time we want to get the curves, returning a new python object.

Note the difference with the network `get_` methods, because the network is mutable. Here the result is immutable, so it makes more sense not to have `get_` methods.

nicolas-pierr replied:
I fixed it. I could not find a good way to make only one java call to get all curves. In Python it is easy to concatenate dataframes, however it is not in java. And there was no easy way to return an array of dataframes without getting into memory allocations troubles. I don't know if I missed an example of this type (passing an array of a generic object from java to python).

sylvlecl commented on pypowsybl/dynamic.py:
```diff
@@ -0,0 +1,190 @@
+        )
+
+    @staticmethod
+    def set_config(path: str, config_name: str) -> None:
```
Should be removed until we support changing the config at runtime.

nicolas-pierr replied:
You are right, it was not meant ot be pushed

sylvlecl reviewed (commented):
A set of comments, did not finish the review yet: thanks for the work so far :)

Commit:
dataframe adders for dynamic mapping

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
renaming quadripole to branch

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
branch side enum

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
new mapping with dataframe, curves name getter

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
quadripole renamed branch
mapping with dataframe
branchSide enum

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
remove omega_ref mapping

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
fix duplicate header from rebase

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
fix rename quadripole to branch

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
tests:rename quadripole to branch,remove omega_ref

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

nicolas-pierr commented:
I rebased on main, I saw that the omega_ref is automaticaly mapped to every synchronized generator. Do you confirm that the user will never have to map omega_ref ?

Commit:
remove curvemapping storing curves names

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

nicolas-pierr commented:
Also I made a branch on pypowsyb-notebooks with an example to run.
All data asssociated (config, network and .par files) is provided.
Specific branch is [here](https://github.com/powsybl/pypowsybl-notebooks/tree/dynawaltz)

Commit:
refactor

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
added some minimal testing

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
load_from_string clarification

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

geofjamg committed:
Merge branch 'main' into dynawaltz_integration

# Conflicts:
#	cpp/src/pypowsybl.h

Commit:
Merge branch 'main' of github.com:powsybl/pypowsybl into main

Commit:
documentation

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
remove omega ref comment

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
use autosummary

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
Merge branch 'main' into dynawaltz_integration

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
Merge remote-tracking branch 'origin/main' into dynawaltz_integration

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
missing }

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

EtienneLt committed:
Merge branch 'main' into dynawaltz_integration

Commit:
Merge remote-tracking branch 'origin/main' into dynawaltz_integration

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

EtienneLt commented on java/src/main/java/com/powsybl/dataframe/dynamic/adders/GeneratorSynchronousFourWindingsAdder.java:
```diff
@@ -0,0 +1,11 @@
+package com.powsybl.dataframe.dynamic.adders;
+
```
headers and author missing

EtienneLt commented on java/src/main/java/com/powsybl/dataframe/dynamic/adders/GeneratorSynchronousFourWindingsProportionalRegulationsAdder.java:
```diff
@@ -0,0 +1,11 @@
+package com.powsybl.dataframe.dynamic.adders;
+
```
headers and author missing

EtienneLt commented on java/src/main/java/com/powsybl/dataframe/dynamic/adders/GeneratorSynchronousThreeWindingsAdder.java:
```diff
@@ -0,0 +1,11 @@
+package com.powsybl.dataframe.dynamic.adders;
+
```
same

EtienneLt commented on java/src/main/java/com/powsybl/dataframe/dynamic/adders/GeneratorSynchronousThreeWindingsProportionalRegulationsAdder.java:
```diff
@@ -0,0 +1,11 @@
+package com.powsybl.dataframe.dynamic.adders;
+
```
same

EtienneLt commented on java/src/main/java/com/powsybl/dataframe/dynamic/adders/OneTransformerLoadAdder.java:
```diff
@@ -0,0 +1,10 @@
+package com.powsybl.dataframe.dynamic.adders;
+
```
same

EtienneLt commented on java/src/main/java/com/powsybl/python/dynamic/CurveMappingSupplier.java:
```diff
@@ -0,0 +1,44 @@
+import com.powsybl.dynamicsimulation.CurvesSupplier;
+import com.powsybl.dynawaltz.DynaWaltzCurve;
+import com.powsybl.iidm.network.Network;
+
```
author missing

EtienneLt commented on java/src/main/java/com/powsybl/python/dynamic/DynamicModelMapper.java:
```diff
@@ -0,0 +1,79 @@
+import com.powsybl.dynawaltz.dynamicmodels.LoadOneTransformer;
+import com.powsybl.iidm.network.Branch;
+import com.powsybl.iidm.network.Network;
+
```
author missing

EtienneLt commented on java/src/main/java/com/powsybl/python/dynamic/DynamicSimulationCFunctions.java:
```diff
@@ -0,0 +1,209 @@
+import com.powsybl.python.network.NetworkCFunctions;
+import com.powsybl.timeseries.DoublePoint;
+import com.powsybl.timeseries.TimeSeries;
+
```
author missing

EtienneLt commented on java/src/main/java/com/powsybl/python/dynamic/DynamicSimulationContext.java:
```diff
@@ -0,0 +1,58 @@
+import com.powsybl.dynamicsimulation.DynamicSimulationResult;
+import com.powsybl.dynamicsimulation.EventModelsSupplier;
+import com.powsybl.iidm.network.Network;
+
```
same

EtienneLt commented on java/src/main/java/com/powsybl/python/dynamic/EventSupplier.java:
```diff
@@ -0,0 +1,51 @@
+import com.powsybl.dynawaltz.events.EventQuadripoleDisconnection;
+import com.powsybl.dynawaltz.events.EventSetPointBoolean;
+import com.powsybl.iidm.network.Network;
+
```
same

Commit:
author and copyrights

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

geofjamg committed:
Merge branch 'main' into dynawaltz_integration

# Conflicts:
#	cpp/src/bindings.cpp

nicolas-pierr commented:
Hello, some changes from this morning broke my compilation. I see that the generators models have changed completly. Can you confirm that GeneratorSynchronous and GeneratorFictious are the new mapping for generators that must be done ? And that GeneratorSynchronousThreeWindings and its friends were deleted from powsybl-dynawaltz ?

flo-dup commented:
Indeed, all synchronous generators have been merged into one `GeneratorSynchronous` class, which takes the library name as new parameter. Therefore you have to change [these mappings](https://github.com/powsybl/pypowsybl/blob/f4a98d1d0bdb9e93be535474713fa719e9db4e58/java/src/main/java/com/powsybl/dataframe/dynamic/adders/DynamicMappingAdderFactory.java#L22-L29) so that they each give a `GeneratorSynchronousAdder` which will retain the correct corresponding library name.

Commit:
refactor all generatorsSynchronous mappings
Added generatorsSynchronous adder

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

nicolas-pierr commented:
I updated the mappings to `GeneratorSynchronous` variants already implemented. They now use the `GeneratorSynchronous` class. I also created a `DynamicMappingAdder` for `GeneratorSynchronous`, it allows to add via dataframe different types of `GeneratorSynchronous` with a single call of python.

Commit:
update dynawaltz events bindings

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
Merge remote-tracking branch 'origin/main' into dynawaltz_integration

nicolas-pierr commented:
I also updated the events mapping with the new eventTime and their arguments.
There are compilations errors due to commits on main.

Commit:
Merge branch 'main' into dynawaltz_integration

Commit:
fix all weird main merges + moved createSeriesMetadata

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

Commit:
fix linting, renamed dynamic_param to parameter_set_id

Signed-off-by: Nicolas PIERRE <nicolas.pierre@artelys.com>

nicolas-pierr commented:
Some information about some changes under the hood:
- I moved a Dataframe function `createSeriesMetadata` from NetworkCUtil to CTypeUtil, as it is not specific to Network and i needed it to add dynawaltz mapping with dataframes.
- I also modified some typings in `dataframes.py` as I called them with enum and not standard python types. And added `_add_index_to_kwargs` that creates an explicit index for a kwargs that will be used as Dataframe.

EtienneLt committed:
Merge branch 'main' into dynawaltz_integration
